Platform/CoffeelakeBoardPkg/Library/Stage2BoardInitLib/Stage2BoardInitLib.c

```diff
@@ -1602,48 +1602,53 @@ SaveNvsData (
   if (EFI_ERROR(Status)) {
     return EFI_NOT_FOUND;
   }
-  if (*(UINT32 *)Address == 0xFFFFFFFF) {
-    Status = SpiConstructor ();
-    if (EFI_ERROR (Status)) {
-      return Status;
-    }
 
-    Status = SpiGetRegionAddress (FlashRegionAll, NULL,  &TotalSize);
-    if (EFI_ERROR (Status)) {
-      return Status;
-    }
+  //
+  // Compare input data against the stored MRC training data
+  // if they match, no need to update again.
+  //
+  if (CompareMem ((VOID *)Address, Buffer, Length) == 0){
+    return EFI_ALREADY_STARTED;
+  }
 
-    Status = SpiGetRegionAddress (FlashRegionBios, &BaseAddress, &RegionSize);
-    if (EFI_ERROR (Status)) {
-      return Status;
-    }
+  Status = SpiConstructor ();
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
 
-    BaseAddress += ((UINT32)(~TotalSize) + 1);
-    if (Address < BaseAddress) {
-      return EFI_ACCESS_DENIED;
-    }
+  Status = SpiGetRegionAddress (FlashRegionAll, NULL,  &TotalSize);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
 
-    Address -= BaseAddress;
-    if ((Address + ROUNDED_UP(Length, KB_(4))) > RegionSize) {
-      return EFI_OUT_OF_RESOURCES;
-    }
+  Status = SpiGetRegionAddress (FlashRegionBios, &BaseAddress, &RegionSize);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
 
+  BaseAddress += ((UINT32)(~TotalSize) + 1);
+  if (Address < BaseAddress) {
```

Comment:
Using " BaseAddress = ((UINT32)(~RegionSize) + 1);", it is not necessary to get TotalSize.

Reply:
Updated new patch